File: geotrellis-python/geotrellis/spark/TileLayerMetadata.py
from geotrellis.spark.KeyBounds import Bounds, EmptyBounds, KeyBounds
from geotrellis.spark.SpatialKey import SpatialKey
from geotrellis.raster.CellSize import CellSize
from geotrellis.vector.Extent import Extent, ProjectedExtent
from geotrellis.python.crs.CRSWrapper import CRS
from geotrellis.spark.tiling.LayoutDefinition import LayoutDefinition

class TileLayerMetadata(object):
    def __init__(self, cellType, layout, extent, crs, bounds):
        self.cellType = cellType
        self.layout = layout
        self.extent = extent
        self.crs = crs
        self.bounds = bounds

    def __eq__(self, other):
        if not isinstance(other, TileLayerMetadata):
            return False
        return (self.cellType == other.cellType and
                self.layout == other.layout and
                self.extent == other.extent and
                self.crs == other.crs and
                self.bounds == other.bounds)

    def __hash__(self):
        return hash((self.cellType, self.layout, self.extent, self.crs, self.bounds))

    def _copy(self, cellType = None, layout = None, extent = None, crs = None, bounds = None):
        cellType = cellType if cellType is not None else self.cellType
        layout = layout if layout is not None else self.layout
        extent = extent if extent is not None else self.extent
        crs = crs if crs is not None else self.crs
        bounds = bounds if bounds is not None else self.bounds
        return TileLayerMetadata(cellType, layout, extent, crs, bounds)

    @property
    def mapTransform(self):
        return self.layout.mapTransform

    @property
    def tileLayout(self):
        return self.layout.tileLayout

    @property
    def layoutExtent(self):
        return self.layout.extent

    @property
    def gridBounds(self):
        return self.mapTransform(self.extent)

    def combine(self, other):
        combinedExtent = self.extent.combine(other.extent)
        combinedLayoutExtent = self.layoutExtent.combine(other.layoutExtent)
        combinedTileLayout = self.tileLayout.combine(other.tileLayout)
        newLayout = LayoutDefinition(
                combinedLayoutExtent,
                combinedTileLayout)

        return self._copy(
                extent = combinedExtent,
                layout = newLayout)

    def updateBounds(self, newBounds):
        if newBounds is EmptyBounds:
            newExtent = Extent(
                    self.extent.xmin, self.extent.ymin,
                    self.extent.xmin, self.extent.ymin)
            return self._copy(
                    extent = newExtent,
                    bounds = newBounds)
        kbExtent = self.mapTransform(newBounds.toGridBounds)
        intersection = kbExtent.intersection(self.extent)
        if intersection is not None:
            return self._copy(bounds = newBounds, extent = intersection)
        else:
            newExtent = Extent(
                    self.extent.xmin, self.extent.ymin,
                    self.extent.xmin, self.extent.ymin)
            return self._copy(bounds = newBounds, extent = newExtent)

    def getComponent(self, componentType):
        if componentType == Extent:
            return self.extent
        elif componentType == CRS:
            return self.crs
        elif componentType == LayoutDefinition:
            return self.layout
        elif componentType == Bounds:
            return self.bounds
        else:
            raise Exception("wrong componentType: {0}".format(componentType))

    def setComponent(self, value):
        if isinstance(value, LayoutDefinition):
            return self._copy(layout = value)
        elif isinstance(value, Bounds):
            return self._copy(bounds = value)
        else:
            raise Exception("wrong argument: {0}".format(value))

    @staticmethod
    def fromRdd(rdd, first, second = None):
        if (isinstance(first, CRS) and
                isinstance(second, LayoutDefinition)):
            crs, layout = first, second
            extent, cellType, _cellSize, bounds = _collectMetadata(rdd)
            kb = bounds.setSpatialBounds(KeyBounds(layout.mapTransform(extent)))
            return TileLayerMetadata(cellType, layout, extent, crs, kb)
        elif isinstance(first, LayoutDefinition) and second is None:
            layout = first
            extent, cellType, _cellSize, bounds, crs = _collectMetadataWithCRS(rdd)
            kb = bounds.setSpatialBounds(KeyBounds(layout.mapTransform(extent)))
            return TileLayerMetadata(cellType, layout, extent, crs, kb)
        else:
            # TODO implement LayoutScheme case
            raise Exception("Not implemented")


def _collectMetadata(rdd):
    def mapper(t):
        key, grid = t
        extent = key.extent
        boundsKey = key.translate(SpatialKey(0,0))
        return (extent,
                grid.cellType,
                CellSize(extent, grid.cols, grid.rows)
                KeyBounds(boundsKey, boundsKey))
    def reducer(tuple1, tuple2):
        extent1, cellType1, cellSize1, bounds1 = tuple1
        extent2, cellType2, cellSize2, bounds2 = tuple2
        return (extent1.combine(extent2),
                cellType1.union(cellType2),
                cellSize1 if cellSize1.resolution < cellSize2.resolution else cellSize2,
                bounds1.combine(bounds2))
    return rdd.map(mapper).reduce(reducer)

def _collectMetadataWithCRS(rdd):
    def mapper(t):
        key, grid = t
        projectedExtent = key.getComponent(ProjectedExtent)
        extent = projectedExtent.extent
        crs = projectedExtent.crs
        boundsKey = key.translate(SpatialKey(0,0))
        return (extent,
                grid.cellType,
                CellSize(extent, grid.cols, grid.rows),
                set(crs),
                KeyBounds(boundsKey, boundsKey))
    def reducer(tuple1, tuple2):
        extent1, cellType1, cellSize1, crs1, bounds1 = tuple1
        extent2, cellType2, cellSize2, crs2, bounds2 = tuple2
        return (extent1.combine(extent2),
                cellType1.union(cellType2),
                cellSize1 if cellSize1.resolution < cellSize2.resolution else cellSize2,
                crs1.union(crs2),
                bounds1.combine(bounds2))
    extent, cellType, cellSize, crsSet, bounds = rdd.map(mapper).reduce(reducer)
    if len(crsSet) != 1:
        raise Exception("Multiple CRS tags found: {0}".format(crsSet))
    return (extent, cellType, cellSize, bounds, crsSet[0])

File: spark/src/main/scala/geotrellis/spark/TileLayerMetadata.scala
/*
 * Copyright 2016 Azavea
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package geotrellis.spark

import geotrellis.proj4.CRS
import geotrellis.raster._
import geotrellis.tiling._
import geotrellis.spark.ingest._
import geotrellis.spark.tiling._
import geotrellis.util._
import geotrellis.vector.{ProjectedExtent, Extent}

import cats.Functor
import cats.implicits._
import org.apache.spark.rdd._

/**
 * @param cellType    value type of each cell
 * @param layout      definition of the tiled raster layout
 * @param extent      Extent covering the source data
 * @param crs         CRS of the raster projection
 */
case class TileLayerMetadata[K](
  cellType: CellType,
  layout: LayoutDefinition,
  extent: Extent,
  crs: CRS,
  bounds: Bounds[K]
) {
  /** Transformations between tiling scheme and map references */
  def mapTransform: MapKeyTransform = layout.mapTransform
  /** TileLayout of the layout */
  def tileLayout: TileLayout = layout.tileLayout
  /** Full extent of the layout */
  def layoutExtent: Extent = layout.extent
  /** GridBounds of data tiles in the layout */
  def tileBounds: TileBounds = mapTransform(extent)

  def combine(other: TileLayerMetadata[K])(implicit b: Boundable[K]): TileLayerMetadata[K] = {
    val combinedExtent       = extent combine other.extent
    val combinedLayoutExtent = layout.extent combine other.layout.extent
    val combinedTileLayout   = layout.tileLayout combine other.layout.tileLayout
    val combinedBounds       = bounds combine other.bounds

    this
      .copy(
        extent = combinedExtent,
        bounds = combinedBounds,
        layout = this.layout
          .copy(
            extent     = combinedLayoutExtent,
            tileLayout = combinedTileLayout
          )
      )
  }

  def updateBounds(newBounds: Bounds[K])(implicit c: Component[K, SpatialKey]): TileLayerMetadata[K] =
    newBounds match {
      case kb: KeyBounds[K] => {
        val SpatialKey(minCol, minRow) = kb.minKey.getComponent[SpatialKey]
        val SpatialKey(maxCol, maxRow) = kb.maxKey.getComponent[SpatialKey]
        val kbExtent = mapTransform(GridBounds(minCol, minRow, maxCol, maxRow))

        kbExtent.intersection(extent) match {
          case Some(e) =>
            copy(bounds = newBounds, extent = e)
          case None =>
            copy(bounds = newBounds, extent = Extent(extent.xmin, extent.ymin, extent.xmin, extent.ymin))
        }
      }
      case EmptyBounds =>
        copy(bounds = newBounds, extent = Extent(extent.xmin, extent.ymin, extent.xmin, extent.ymin))
    }
}

object TileLayerMetadata {
  implicit def toLayoutDefinition(md: TileLayerMetadata[_]): LayoutDefinition =
    md.layout

  implicit def extentComponent[K]: GetComponent[TileLayerMetadata[K], Extent] =
    GetComponent(_.extent)

  implicit def crsComponent[K]: GetComponent[TileLayerMetadata[K], CRS] =
    GetComponent(_.crs)

  implicit def layoutComponent[K: SpatialComponent]: Component[TileLayerMetadata[K], LayoutDefinition] =
    Component(_.layout, (md, l) => md.copy(layout = l))

  implicit def boundsComponent[K: SpatialComponent]: Component[TileLayerMetadata[K], Bounds[K]] =
    Component(_.bounds, (md, b) => md.updateBounds(b))

  implicit def mergable[K: Boundable]: merge.Mergable[TileLayerMetadata[K]] =
    new merge.Mergable[TileLayerMetadata[K]] {
      def merge(t1: TileLayerMetadata[K], t2: TileLayerMetadata[K]): TileLayerMetadata[K] =
        t1.combine(t2)
    }

  implicit val tileLayerMetadataFunctor: Functor[TileLayerMetadata] = new Functor[TileLayerMetadata] {
    def map[A, B](fa: TileLayerMetadata[A])(f: A => B): TileLayerMetadata[B] =
      fa.copy(bounds = fa.bounds.map(f))
  }

  private def collectMetadata[
    K: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)]): (Extent, CellType, CellSize, KeyBounds[K2]) = {
    rdd
      .map { case (key, grid) =>
        val extent = key.extent
        val boundsKey = key.translate(SpatialKey(0,0))
        // Bounds are return to set the non-spatial dimensions of the keybounds; the spatial keybounds are set outside this call.
        (extent, grid.cellType, CellSize(extent, grid.cols, grid.rows), KeyBounds(boundsKey, boundsKey))
      }
      .reduce { (tuple1, tuple2) =>
        val (extent1, cellType1, cellSize1, bounds1) = tuple1
        val (extent2, cellType2, cellSize2, bounds2) = tuple2
        (
          extent1.combine(extent2),
          cellType1.union(cellType2),
          if (cellSize1.resolution < cellSize2.resolution) cellSize1 else cellSize2,
          bounds1.combine(bounds2)
        )
      }
  }

  private def collectMetadataWithCRS[
    K: GetComponent[?, ProjectedExtent]: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)]): (Extent, CellType, CellSize, KeyBounds[K2], CRS) = {
    val (extent, cellType, cellSize, crsSet, bounds) =
      rdd
      .map { case (key, grid) =>
        val ProjectedExtent(extent, crs) = key.getComponent[ProjectedExtent]
        val boundsKey = key.translate(SpatialKey(0,0))
        // Bounds are return to set the non-spatial dimensions of the keybounds; the spatial keybounds are set outside this call.
        (extent, grid.cellType, CellSize(extent, grid.cols, grid.rows), Set(crs), KeyBounds(boundsKey, boundsKey))
      }
      .reduce { (tuple1, tuple2) =>
        val (extent1, cellType1, cellSize1, crs1, bounds1) = tuple1
        val (extent2, cellType2, cellSize2, crs2, bounds2) = tuple2
        (
          extent1.combine(extent2),
          cellType1.union(cellType2),
          if (cellSize1.resolution < cellSize2.resolution) cellSize1 else cellSize2,
          crs1 ++ crs2,
          bounds1.combine(bounds2)
          )
      }
    require(crsSet.size == 1, s"Multiple CRS tags found: $crsSet")
    (extent, cellType, cellSize, bounds, crsSet.head)
  }

  /**
    * Compose Extents from given raster tiles and fit it on given
    * TileLayout.
    */
  def fromRDD[
    K: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], crs: CRS, layout: LayoutDefinition): TileLayerMetadata[K2] = {
    val (extent, cellType, _, bounds) = collectMetadata(rdd)
    val kb = bounds.setSpatialBounds(KeyBounds(layout.mapTransform(extent)))
    TileLayerMetadata(cellType, layout, extent, crs, kb)
  }

  /**
    * Compose Extents from given raster tiles and use LayoutScheme to
    * create the LayoutDefinition.
    */
  def fromRDD[
    K: (? => TilerKeyMethods[K, K2]) ,
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], crs: CRS, scheme: LayoutScheme): (Int, TileLayerMetadata[K2]) = {
    val (extent, cellType, cellSize, bounds) = collectMetadata(rdd)
    val LayoutLevel(zoom, layout) = scheme.levelFor(extent, cellSize)
    val kb = bounds.setSpatialBounds(KeyBounds(layout.mapTransform(extent)))
    (zoom, TileLayerMetadata(cellType, layout, extent, crs, kb))
  }

  /**
    * Compose Extents from given raster tiles and use
    * [[geotrellis.spark.tiling.ZoomedLayoutScheme]] to create the
    * [[geotrellis.spark.tiling.LayoutDefinition]].
    */
  def fromRDD[
    K: (? => TilerKeyMethods[K, K2]) ,
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], crs: CRS, scheme: ZoomedLayoutScheme):
    (Int, TileLayerMetadata[K2]) =
      _fromRDD[K, V, K2](rdd, crs, scheme, None)

  /**
    * Compose Extents from given raster tiles using
    * [[geotrellis.spark.tiling.ZoomedLayoutScheme]] and a maximum
    * zoom value.
    */
  def fromRDD[
    K: (? => TilerKeyMethods[K, K2]) ,
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], crs: CRS, scheme: ZoomedLayoutScheme, maxZoom: Int):
    (Int, TileLayerMetadata[K2]) =
      _fromRDD[K, V, K2](rdd, crs, scheme, Some(maxZoom))

  private def _fromRDD[
    K: (? => TilerKeyMethods[K, K2]) ,
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], crs: CRS, scheme: ZoomedLayoutScheme, maxZoom: Option[Int]):
    (Int, TileLayerMetadata[K2]) = {
      val (extent, cellType, cellSize, bounds) = collectMetadata(rdd)
      val LayoutLevel(zoom, layout) = maxZoom match {
        case Some(zoom) => scheme.levelForZoom(maxZoom.get)
        case _ => scheme.levelFor(extent, cellSize)
      }
    val kb = bounds.setSpatialBounds(KeyBounds(layout.mapTransform(extent)))
    (zoom, TileLayerMetadata(cellType, layout, extent, crs, kb))
  }

  def fromRDD[
    K: GetComponent[?, ProjectedExtent]: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], scheme: LayoutScheme): (Int, TileLayerMetadata[K2]) = {
    val (extent, cellType, cellSize, bounds, crs) = collectMetadataWithCRS(rdd)
    val LayoutLevel(zoom, layout) = scheme.levelFor(extent, cellSize)
    val kb = bounds.setSpatialBounds(KeyBounds(layout.mapTransform(extent)))
    (zoom, TileLayerMetadata(cellType, layout, extent, crs, kb))
  }

  def fromRDD[
    K: GetComponent[?, ProjectedExtent]: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)],  scheme: ZoomedLayoutScheme):
    (Int, TileLayerMetadata[K2]) =
      _fromRDD[K, V, K2](rdd, scheme, None)

  def fromRDD[
    K: GetComponent[?, ProjectedExtent]: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], scheme: ZoomedLayoutScheme, maxZoom: Int):
    (Int, TileLayerMetadata[K2]) =
      _fromRDD[K, V, K2](rdd, scheme, Some(maxZoom))

  private def _fromRDD[
    K: GetComponent[?, ProjectedExtent]: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], scheme: ZoomedLayoutScheme, maxZoom: Option[Int]):
  (Int, TileLayerMetadata[K2]) = {
    val (extent, cellType, cellSize, bounds, crs) = collectMetadataWithCRS(rdd)
    val LayoutLevel(zoom, layout) = maxZoom match {
      case Some(zoom) => scheme.levelForZoom(maxZoom.get)
      case _ => scheme.levelFor(extent, cellSize)
    }
    val kb = bounds.setSpatialBounds(KeyBounds(layout.mapTransform(extent)))
    (zoom, TileLayerMetadata(cellType, layout, extent, crs, kb))
  }

  def fromRDD[
    K: GetComponent[?, ProjectedExtent]: (? => TilerKeyMethods[K, K2]),
    V <: CellGrid[Int],
    K2: SpatialComponent: Boundable
  ](rdd: RDD[(K, V)], layoutDefinition: LayoutDefinition): TileLayerMetadata[K2] = {
    val (extent, cellType, cellSize, bounds, crs) = collectMetadataWithCRS(rdd)
    val kb = bounds.setSpatialBounds(KeyBounds(layoutDefinition.mapTransform(extent)))
    TileLayerMetadata(cellType, layoutDefinition, extent, crs, kb)
  }
}
